mqtt_as/mqtt_as.py

```diff
@@ -317,8 +317,9 @@ async def broker_up(self):  # Test broker connectivity
 
     async def disconnect(self):
         try:
-            async with self.lock:
-                self._sock.write(b"\xe0\0")
+            if self._sock is not None:
+                async with self.lock:
+                        self._sock.write(b"\xe0\0")
```

Comment:
To retain the original logic the socket should be closed here.

I still prefer my version where the try-except block is only run if `self._sock` exists. I also wonder if we should pause between writing the close command and closing the socket:
```python
    async def disconnect(self):
        if self._sock is not None:
            try:
                async with self.lock:
                    self._sock.write(b"\xe0\0")
                    await asyncio.sleep_ms(100)  # ? Do you think this is wise?
            except OSError:
                pass
            self._sock.close()  # Retain original logic
        self._has_connected = False
```

Reply:
Adding the sleep is a sold plan. I think could be the cause of an issue I have been looking for (had not started full investigation yet). In our device overview dashboard, we have seen a larger than expected number of devices time out their MQTT connection (up to 50% of the fleet sometimes). 

The `socket.close()` documentation does say the following:
>Mark the socket closed. Once that happens, all future operations on the socket object will fail. The remote end will >receive no more data (after queued data is flushed).

However, taking a brief look at the MP source code, I am uncertain if this is happening for all ports. I could be wrong here, as have only scanned quickly through, maybe there is some other system responsible for this that I overlooked. 

Lacking a `socket.flush()` method, this seems like a good way to give the system a little breathing room to get that message out.

Reply:
As for your other remark, regarding keeping the original logic I would like to suggest the following:

```python
    async def disconnect(self):
        if self._sock is not None:
            try:
                async with self.lock:
                    self._sock.write(b"\xe0\0")
                    await asyncio.sleep_ms(100)
            except OSError:
                pass
            self._close()  # Use ._close here instead of sock.close()
        self._has_connected = False
```

This way, if there is any every special logic for closing a socket on a specific platform, no changes are needed to the disconnect function.

Reply:
I'll also investigate `socket.setblocking` options to improve disconnect stability, but that is something for outside this PR.